controller/src/scheduler.rs

```diff
@@ -0,0 +1,107 @@
+use cron::Schedule;
+use snafu::ResultExt;
+use std::env;
+use std::str::FromStr;
+
+// Defines the cron expression scheduler related env variable names
+const SCHEDULER_CRON_SECONDS_ENV_VAR: &str = "SCHEDULER_CRON_SECONDS";
+const SCHEDULER_CRON_MINUTES_ENV_VAR: &str = "SCHEDULER_CRON_MINUTES";
+const SCHEDULER_CRON_HOURS_ENV_VAR: &str = "SCHEDULER_CRON_HOURS";
+const SCHEDULER_CRON_DAY_OF_MONTH_ENV_VAR: &str = "SCHEDULER_CRON_DAY_OF_MONTH";
+const SCHEDULER_CRON_MONTH_ENV_VAR: &str = "SCHEDULER_CRON_MONTH";
+const SCHEDULER_CRON_DAY_OF_WEEK_ENV_VAR: &str = "SCHEDULER_CRON_DAY_OF_WEEK";
+const SCHEDULER_CRON_YEARS_ENV_VAR: &str = "SCHEDULER_CRON_YEARS";
+
+/// The module-wide result type.
+type Result<T> = std::result::Result<T, scheduler_error::Error>;
+
+struct BrupopCronScheduler {
+    seconds: String,
+    minutes: String,
+    hours: String,
+    day_of_month: String,
+    month: String,
+    day_of_week: String,
+    year: String,
+}
+
+impl BrupopCronScheduler {
+    pub fn new() -> Self {
+        let scheduler_cron_seconds = read_env_var(SCHEDULER_CRON_SECONDS_ENV_VAR).unwrap();
+        let scheduler_cron_minutes = read_env_var(SCHEDULER_CRON_MINUTES_ENV_VAR).unwrap();
+        let scheduler_cron_hours = read_env_var(SCHEDULER_CRON_HOURS_ENV_VAR).unwrap();
+        let scheduler_cron_day_of_month =
+            read_env_var(SCHEDULER_CRON_DAY_OF_MONTH_ENV_VAR).unwrap();
+        let scheduler_cron_month = read_env_var(SCHEDULER_CRON_MONTH_ENV_VAR).unwrap();
+        let scheduler_cron_day_of_week = read_env_var(SCHEDULER_CRON_DAY_OF_WEEK_ENV_VAR).unwrap();
+        let scheduler_cron_years = read_env_var(SCHEDULER_CRON_YEARS_ENV_VAR).unwrap();
+
+        BrupopCronScheduler {
+            seconds: scheduler_cron_seconds,
+            minutes: scheduler_cron_minutes,
+            hours: scheduler_cron_hours,
+            day_of_month: scheduler_cron_day_of_month,
+            month: scheduler_cron_month,
+            day_of_week: scheduler_cron_day_of_week,
+            year: scheduler_cron_years,
+        }
+    }
+
+    // generate a valid cron expression format string value based on provided env variables.
+    pub fn generate_cron_expression(&self) -> String {
+        format!(
+            "{} {} {} {} {} {} {}",
+            self.seconds,
+            self.minutes,
+            self.hours,
+            self.day_of_month,
+            self.month,
+            self.day_of_week,
+            self.year
+        )
+    }
+}
+
+// cron expression parser and schedule explorer
+pub async fn generate_schedule() -> Result<Schedule> {
+    let brupop_scheduler = BrupopCronScheduler::new();
+    let cron_expression = brupop_scheduler.generate_cron_expression();
+
+    let schedule = Schedule::from_str(&cron_expression.as_str()).context(
```

Comment:
nit, `&String` should deref-coerce to `&str`:
```suggestion
    let schedule = Schedule::from_str(&cron_expression).context(
```